pallets/parachain-system/src/lib.rs

```diff
@@ -321,25 +302,36 @@ pub mod pallet {
 
 			Self::validate_validation_data(&vfp);
 
+			let relay_state_proof = RelayChainStateProof::new(
+				T::SelfParaId::get(),
+				vfp.relay_parent_storage_root,
+				relay_chain_state,
+			)
+			.expect("Invalid relay chain state proof");
+
 			// initialization logic: we know that this runs exactly once every block,
 			// which means we can put the initialization logic here to remove the
 			// sequencing problem.
-			if let Some(apply_block) = <PendingRelayChainBlockNumber<T>>::get() {
-				if vfp.relay_parent_number >= apply_block {
-					<PendingRelayChainBlockNumber<T>>::kill();
+			let upgrade_go_ahead_signal = relay_state_proof
+				.read_upgrade_go_ahead_signal()
+				.expect("Invalid upgrade go ahead signal");
+			match upgrade_go_ahead_signal {
+				Some(relay_chain::v1::UpgradeGoAhead::GoAhead) => {
 					let validation_code = <PendingValidationCode<T>>::take();
```

Comment:
Shouldn't we double-check that the entry is present in the storage before taking it?

Reply:
Like in the case the relay-chain flashed us with `GoAhead` when none was expected? Hm, theoretically that should never happen. Although I can see how a logic error be it on the relay-chain or the parachain can lead to that. If that happens the parachain is screwed most likely, but perhaps it won't hurt if we try to save the day here.

Reply:
This is not done yet?